### PR TITLE
Bugfix/1352 incorrect report multiple lines

### LINKED ISSUE #1352: Finding sometimes is incorrectly reported to span multiple lines
**Describe the bug**
Sometimes, the reported start and end lines are different even though the end line has nothing to do with the finding

**To Reproduce**
1. Create a new directory containing only a single file (`.env.prod`) with the following content:
```
DB_HOST=example.com
DB_PORT=443
DB_USERNAME=postgres
DB_PASSWORD=8ae31cacf141669ddfb5da
DB_NAME=best_db
DB_SSL=true
```
2. Run `gitleaks detect --no-git --report-path o`
3. In `o` , you will find the following report:
```
[
 {
  "Description": "Detected a Generic API Key, potentially exposing access to various services and sensitive operations.",
  "StartLine": 4,
  "EndLine": 5,
  "StartColumn": 5,
  "EndColumn": 1,
  "Match": "PASSWORD=8ae31cacf141669ddfb5da",
  "Secret": "8ae31cacf141669ddfb5da",
  "File": ".env.prod",
  "SymlinkFile": "",
  "Commit": "",
  "Entropy": 3.5383105,
  "Author": "",
  "Email": "",
  "Date": "",
  "Message": "",
  "Tags": [],
  "RuleID": "generic-api-key",
  "Fingerprint": ".env.prod:generic-api-key:4"
 }
]
```

**Expected behavior**
The start line is correct. However, the start column already is wrong because start column 5 means that the finding is

```
SSWORD=8ae31cacf141669ddfb5da
DB
```

if columns are 0-indexed and

```
ASSWORD=8ae31cacf141669ddfb5da
D
```

if columns are 1-indexed.

However, what really seems very wrong is the jump into line 5 because `DB_NAME` has nothing to do with the password.

**Screenshots**

**Basic Info (please complete the following information):**
 - OS: Manjaro
 - Gitleaks Version: 8.18.2

cc @zricethezav

### PR DESCRIPTION
**Pull request by recreator66**
### Description:
This fixes #1352 
### Checklist:

* [x] Does your PR pass tests?
* [x] Have you written new tests for your changes?
* [x] Have you lint your code locally prior to submission?

### CODE REVIEW

File: detect/location.go [hunk 1]
```diff
@@ -50,6 +50,15 @@ func location(fragment Fragment, matchIndex []int) Location {
...
 			location.endLineIndex = newLineByteIndex
 		}
+
+		// Fixes: https://github.com/gitleaks/gitleaks/issues/1352
+		// fixes the bug that a new line from default secretSuffix
+		// leads in the reporting to a false positive in the following line
+		if location.endColumn == 1 {
+			location.endLine -= 1 // -1 because of newline in secretSuffix
+			location.endColumn = prevNewLine - location.startLineIndex
+		}
```
Comment (rgmz) [inline detect/location.go:60]: Is it possible to add a test for this? Even with the comment, the reason why this is necessary isn't immediately obvious.

Comment (recreator66) [reply detect/location.go:60]: I will try to create a test for this during this week.

Comment (recreator66) [reply detect/location.go:60]: I have added a test case for this issue. In addition it was necessary to adjust the generic.toml within the testdata directory and to change the viper config in the test function.

@rgmz please have a look at the latest commits and let me know if something needs to be adjusted :)

**Comment by recreator66** [comment]
I have added a test case for this issue. In addition it was necessary to adjust the generic.toml within the testdata directory and to change the viper config in the test function.

@rgmz please have a look at the latest commits and let me know if something needs to be adjusted :)

File: detect/detect_test.go [hunk 2]
```diff
@@ -603,11 +603,32 @@ func TestFromFiles(t *testing.T) {
...
+			expectedFindings: []report.Finding{
+				{
+					Description: "Generic API Key",
+					StartLine:   4,
+					EndLine:     4,
+					StartColumn: 5,
+					EndColumn:   35,
+					Match:       "PASSWORD=8ae31cacf141669ddfb5da",
+					Secret:      "8ae31cacf141669ddfb5da",
+					Line:        "\nDB_PASSWORD=8ae31cacf141669ddfb5da\nDB_NAME=best_db",
```
Comment (rgmz) [inline detect/detect_test.go:618]: I _think_ this exposes a deeper issue. The `EndLine` and `EndColumn` are now correct but the `line` information was previously populated with the incorrect coordinates.

\+ @zricethezav Shouldn't `line` be a single line? Or is this expected behaviour when a match spans multiple lines?

**Review by rgmz — COMMENTED** [review]
Great addition! I'd like to get @zricethezav's thoughts on my question, as there may be something else that needs to be changed,

**Comment by rgmz** [comment]
I think the root cause is that newlines are being trimmed here:
https://github.com/gitleaks/gitleaks/blob/ed19c4e8b417313e7ab6878327167c558f5f98f8/detect/detect.go#L266

But the original match indices — including newlines — are being passed to `location`:
https://github.com/gitleaks/gitleaks/blob/ed19c4e8b417313e7ab6878327167c558f5f98f8/detect/location.go#L21-L22

**Comment by recreator66** [comment]
@rgmz I think the root cause starts even earlier..

https://github.com/gitleaks/gitleaks/blob/ed19c4e8b417313e7ab6878327167c558f5f98f8/detect/detect.go#L263

matchIndices already returns the start of the additional line (in our case `DB_NAME=best_db`). This matches also with the result of regex101

regex101
matchIndices

But the end of the secret is actually matchIndex[1]-1 = 87

![matchIndex[1]](https://github.com/user-attachments/assets/5527d00f-3062-4ff2-9e1e-bdd2a2b74847)

An alternative solution would be to review the regex rules and to avoid `\n` and replace them with  hex char`\x20`. However this would only represent a single whitespace and might have a huge impact on the secret detection..

What's your guys opinion on how to proceed? 😄

**Comment by recreator66** [comment]
Since the `secret` already trims `\n`we could easily use it's length to adjust the `matchIndex[1]`. I think it is not necessary to trim whitespaces at the beginning since the regex rules most likely search for keywords at first.

https://github.com/gitleaks/gitleaks/blob/ed19c4e8b417313e7ab6878327167c558f5f98f8/detect/detect.go#L266

I will do a quick commit and see if it works as expected 👍

**Comment by recreator66** [comment]
I have adjusted matchIndex[1] and changed the expected line in the newly created test case accordingly 👍

File: detect/detect_test.go (hunk 2 above)
Comment (zricethezav) [reply detect/detect_test.go]: > + @zricethezav Also shouldn't line be a single line? Or is this expected behaviour when a match spans multiple lines?

Line should be a single line, correct

**Comment by zricethezav** [comment]
Amazing work @recreator66 and @rgmz! wrangling the location code is not an easy task. LGTM, merging

**Pull request merged**